### PR TITLE
Use container elements to switch frames for testdriver rather than indices

### PR DESCRIPTION
**Pull request by jgraham**
This is more robust against concurrent DOM changes.

### CODE REVIEW

File: tools/wptrunner/wptrunner/executors/protocol.py [hunk 1]
```diff
@@ -393,29 +393,32 @@ def switch_to_window(self, wptrunner_id, initial_window=None):
...
-                    # None here makes us switch back to the parent after we've processed the frame
-                    stack.append(None)
-                    stack.append(frame_id)
+                try:
+                    self._switch_to_frame(item)
+                except ValueError:
+                    # The frame no longer exists, so continue
+                    pass
+
+            handle_window_id, nested_context_containers = self.parent.base.execute_script("""
```
Comment (karlt) [inline tools/wptrunner/wptrunner/executors/protocol.py:402]: A `try`/`except` block is [required](https://treeherder.mozilla.org/logviewer?job_id=344749984&repo=try&lineNumber=1118) on the `execute_script()` call to handle concurrent DOM manipulation.

File: tools/wptrunner/wptrunner/executors/protocol.py [hunk 2]
```diff
@@ -393,29 +393,32 @@ def switch_to_window(self, wptrunner_id, initial_window=None):
...
+let contextParents = Array.from(document.querySelectorAll("frame, iframe, embed, object"))
+  .filter(elem => (elem.localName !== "embed" && elem.contentWindow) || elem.getSVGDocument());
```
Comment (karlt) [inline tools/wptrunner/wptrunner/executors/protocol.py:404]: `HTMLFrameElement` [doesn't](https://html.spec.whatwg.org/multipage/obsolete.html#htmlframeelement) have a `getSVGDocument()` method.  I don't know how much we care about frames, but I wonder whether you meant `elem.localName !== "embed" ? elem.contentWindow : elem.getSVGDocument()`?

However, I wonder whether there is any way for `contentWindow` to be `null` on a container in a document with a non-null browsing context.
And [`getSVGDocument()`](https://html.spec.whatwg.org/multipage/embedded-content-other.html#dom-media-getsvgdocument) returns `null` if the SVG document is not same-origin.

The safe approach might be to be more generous here at the possible cost of more iterations in this loop.

File: tools/wptrunner/wptrunner/executors/protocol.py [hunk 3]
```diff
@@ -393,29 +393,46 @@ def switch_to_window(self, wptrunner_id, initial_window=None):
                 if not first or item != initial_window:
                     self.parent.base.set_window(item)
             else:
-                self._switch_to_frame(item)
+                try:
+                    self._switch_to_frame(item)
+                except ValueError:
+                    # The frame no longer exists, or doesn't have a nested browsng context, so continue
+                    pass
```
Comment (karlt) [inline tools/wptrunner/wptrunner/executors/protocol.py:400]: Do you mean s/`pass`/`continue`/?

File: tools/wptrunner/wptrunner/executors/protocol.py [hunk 4]
```diff
@@ -393,29 +393,46 @@ def switch_to_window(self, wptrunner_id, initial_window=None):
...
+            handle_window_id, nested_context_containers = result
+
+            if handle_window_id and str(handle_window_id) == wptrunner_id:
+                return
+
+            for elem in reversed(nested_context_containers):
+                # None here makes us switch back to the parent after we've processed the frame
+                stack.append(None)
+                stack.append(elem)
             first = False
```
Comment (karlt) [inline tools/wptrunner/wptrunner/executors/protocol.py:427]: The new `continue`s skip this line that clears `first`, which can lead to problems if the window has been changed and the next is `initial_window`.  Perhaps it's simplest to clear `first` immediately after `self.parent.base.set_window(item)`.  Or `first` could be cleared immediately before the `execute_script` `try` block, which would in some ways be more clearly correct.

**Review by karlt — APPROVED** [review]

File: tools/wptrunner/wptrunner/executors/protocol.py (hunk 4 above)
Comment (jgraham) [reply tools/wptrunner/wptrunner/executors/protocol.py:427]: Oops. Good catch.

**Pull request merged**